### PR TITLE
[data] feat: Implement multi-directional sort over Ray Data datasets

### PR DESCRIPTION
**Pull request by sumanthratna**
cc @neilshah12 @richardliaw 

## Why are these changes needed?

Allows Ray Data users to sort multiple columns in different directions. 

## Related issue number

n/a

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/util.py [hunk 1]
```diff
@@ -767,7 +767,7 @@ def find_partition_index(
         else:
             left = prevleft + np.searchsorted(col_vals, desired_val, side="left")
             right = prevleft + np.searchsorted(col_vals, desired_val, side="right")
-    return right if descending is True else left
+    return right if descending[i] is True else left
```
Comment (sumanthratna) [inline python/ray/data/_internal/util.py:770]: Does the index on `descending` matter here? If so, is this correct?

File: python/ray/data/_internal/planner/exchange/sort_task_spec.py [hunk 2]
```diff
@@ -209,11 +207,14 @@ def is_na(x):
                 return np.isnan(x)
             return False
 
-        def key_fn_with_nones(sample):
-            return tuple(NULL_SENTINEL if is_na(x) else x for x in sample)
-
-        # Sort the list, but Nones should be NULL_SENTINEL to ensure safe sorting.
-        samples_list = sorted(samples_list, key=key_fn_with_nones)
+        # To allow multi-directional sort, we utilize Python's stable sort:
+        # we sort several times with different directions.
```
Comment (richardliaw) [inline python/ray/data/_internal/planner/exchange/sort_task_spec.py:211]: can you make a comment as to why you go in reverse

File: python/ray/data/_internal/planner/sort.py [hunk 3]
```diff
@@ -53,10 +53,9 @@ def fn(
             )
         else:
             boundaries = [(b,) for b in sort_key.boundaries]
+            if sort_key.get_descending()[0]:
+                boundaries = boundaries[::-1]
```
Comment (richardliaw) [inline python/ray/data/_internal/planner/sort.py:57]: does this actually work? what if there are multiple indices for the boundaries, and the first index is the same throughout, and the boundaries are sorted by the second index?

File: python/ray/data/_internal/util.py [hunk 4]
```diff
@@ -745,7 +745,7 @@ def find_partition_index(
             col_vals[null_mask] = NULL_SENTINEL
 
         prevleft = left
```
Comment (richardliaw) [inline python/ray/data/_internal/util.py:747]: can we write some tests for this method in particular? like test_util.py

Comment (richardliaw) [reply python/ray/data/_internal/util.py:747]: also ... is this even implemented properly in the first place? `desired` is from `boundary` which is in Key space, not a column index (though it is used as a column index...

File: python/ray/data/_internal/planner/sort.py (hunk 3 above)
Comment (sumanthratna) [reply python/ray/data/_internal/planner/sort.py:57]: This if-statement is only reached if the user provided `boundaries`, and we currently restrict user-passed `boundaries` to apply to the primary sort key (so a single boundary element of `boundaries` a "singleton" tuple with only one index)

File: python/ray/data/_internal/util.py (hunk 4 above)
Comment (sumanthratna) [reply python/ray/data/_internal/util.py:747]: > though it is used as a column index...

Not sure I follow. If a user passed `boundaries=[-10, 0, 2, 5]` then `desired` will take on the values `(-10,)`, `(0,)`, `(2,)`, and `(5,)`. Isn't this correct?

Comment (richardliaw) [reply python/ray/data/_internal/util.py:747]: I'm primarily thinking about the code above:

```
for i in range(len(desired)): # L730 - i is 0 to the number of boundaries provided
...
    col_name = columns[i]  # L733 - why are we using i to index columns

Comment (sumanthratna) [reply python/ray/data/_internal/util.py:747]: if you trace back `columns`, it comes from `sort_key.get_columns()` which returns the columns passed to `ds.sort`—i.e., `columns` is the list of columns to be sorted. 

This is also why indexing `descending[i]` is okay. We have `len(columns) == len(boundaries) == len(descending)`

Comment (richardliaw) [reply python/ray/data/_internal/util.py:747]: why do we assume `We have len(columns) == len(boundaries)`? Can't boundaries be arbitrary?

Comment (sumanthratna) [reply python/ray/data/_internal/util.py:747]: yes that is definitely correct and I don't know why I said that

Comment (sumanthratna) [reply python/ray/data/_internal/util.py:747]: resolved offline: not an issue

TODO for these two functions in `_util`: 
- update types
- add documentation
- add unit tests (especially to verify that the `partition[i]` in `return` is correct

**Comment by sumanthratna** [comment]
@richardliaw and I discussed offline; we'll add more documentation and add unit tests for the partition functions. In the future, we'd like to consider removing the user-provided `boundaries` argument, and instead add a new method called `partition_by_key` to Datasets

File: python/ray/data/_internal/planner/sort.py (hunk 3 above)
Comment (sumanthratna) [reply python/ray/data/_internal/planner/sort.py:57]: resolved offline: not an issue

TODO for this if-statement: 
- add documentation

File: python/ray/data/_internal/util.py [hunk 5]
```diff
@@ -720,9 +720,17 @@ def unify_block_metadata_schema(
 
 def find_partition_index(
     table: Union["pyarrow.Table", "pandas.DataFrame"],
-    desired: List[Any],
+    desired: Tuple[Any],
```
Comment (richardliaw) [inline python/ray/data/_internal/util.py:723]: should this be KeyType instead?

File: python/ray/data/_internal/util.py [hunk 6]
```diff
@@ -720,9 +720,17 @@ def unify_block_metadata_schema(
...
     sort_key: "SortKey",
 ) -> int:
+    """For the given block, find the index where the desired value should be
+    added, to maintain sorted order.
+
+    We do this by iterating over each column, starting with the primary sort key,
+    and binary searching for the desired value in the column. Each binary search
+    shortens the "range" of indices (represented by ``left`` and ``right``, which
+    are indices of rows) where the desired value could be inserted.
+    """
```
Comment (richardliaw) [inline python/ray/data/_internal/util.py:733]: add args: and returns: like in google docstring?

File: python/ray/data/_internal/util.py [hunk 7]
```diff
@@ -767,10 +781,14 @@ def find_partition_index(
...
             left = prevleft + np.searchsorted(col_vals, desired_val, side="left")
             right = prevleft + np.searchsorted(col_vals, desired_val, side="right")
-    return right if descending is True else left
+    return right if descending[0] is True else left
 
 
-def find_partitions(table, boundaries, sort_key):
+def find_partitions(
+    table: Union["pyarrow.Table", "pandas.DataFrame"],
+    boundaries: List[Tuple[Any]],
```
Comment (richardliaw) [inline python/ray/data/_internal/util.py:789]: List[KeyType] i think

**Review by richardliaw — APPROVED** [review]

**Pull request merged**